### DISCUSSION TITLE
How to build dataset for node regression through fusing multiple graph

### DISCUSSION THREAD

**Discussion by xyqbest** [pyg-team/pytorch_geometric#3264] [Q&A] [answered]
My dataset is generated by lots of sensors (assumpting 100 sensors) from time step 1 to time step 60. Each sensor generates 10 features and is linked by another sensor according to some relationship such as domain knowledge. For each time step, some sensors do not work and do not generate data (it could be fill with Na or 0).

My GNN model is to predict the node states (10 features) in the kth time step through fusing the data of the previous time steps (e.g., k-1, k-2, k-3). It means that the column number of X is 30 and the column number of Y is 10.

Finally, I should get 57 graphs, each graph contains the fusion of three previous graphs as X and one graph as Y.

When I build my dataset, I have several questions:

❓ Question 1
Need I number the node from 0 to 99 in the edge_index of torch_geometric.data.Data and drop their labels?

❓ Question 2
For example, I have an edge list :
B-A
B-C
C-D
A-C
Should I keep the map order (for example, {B:0,A:1,C:2,D:3}) in all 57 graphs?

❓ Question 3
Should the row order in the X of torch_geometric.data.Data be from 0 to 99 respectively? Or can the row order be random?

❓ Question 4
In each step, some nodes do not contain any information, should I remove them and reorder the node number?

For example, in the kth time step, only 80 sensors are working, should I remove the 20 non-working nodes and reorder the node number from 0 to 79? But this could lead to the row number of X is uncertain, which could cause the dimension error during the running program.

Thank you for your kind and patient answers!

**Comment by rusty1s** [comment, marked as answer]
1. Yes, nodes in your graph should be labeled from `0` to `num_nodes - 1`, so that, e.g., the node feature `x[10]` corresponds to index values `10` in your `edge_index`.
2. Yes, i think this is a requirement in your setup, as you are operating across different timestamps.
3. Yes, as mentioned above.
4. I think this is up to you. In general, simply "dropping" them by setting their node features to zero results in probably the most easiest implementation. Furthermore, if some nodes do not have any information, they can still receive some from surrounding nodes.

**Reply by xyqbest** [reply]
:smiley: Thanks for your detailed answer. I have several new questions

As you commented, “…if some nodes do not have any information, they can still receive some from surrounding nodes.” If some sensors only work at the first time step and then do not work anymore. It seems these nodes are not necessary. However, your suggestion means that they should be included in the remaining time steps.

❓Question 1:
If I have another graph dataset including 200 sensors, I need to train another network and can not apply the current GNN model to a new graph dataset?

❓Question 2:
Can some methods train on a graph dataset and predict another graph dataset including different nodes.

❓Question 3:
How to split the training dataset and test dataset? If all nodes must be included, it only seems to split the training dataset and test dataset by time steps?

**Reply by rusty1s** [reply]
1. Your GNN model works independent of the number of nodes, as those will share weights. As a result, you will be able to include a new dataset with less or more sensors.
2. Yes (as long as features have the same dimensionality)
3. You can either split by sensors, timestamps or dataset (e.g., a new environment of sensors). I think in your scenario it makes the most sense to split by time, e.g. use the last 10 timestamps for evaluation.

**Reply by xyqbest** [reply]
If my GNN model applies a new dataset with fewer or more sensors, is the edge number not matter?
For example, node “1” in graph A has 7 edges. I train a GNN based on graph A. Then I apply the GNN to graph B but node “1” in graph B has 20 edges or 3 edges. 
Is there any problem with the aggregation of surrounding nodes at node “1” in graph B? For example, the error increases?

😃 Thanks again!

**Reply by rusty1s** [reply]
The GNN model should be able to correctly execute. Its generalization capabilities need to be verified experimentally though :)
